packages/pagination/src/pagination.js

```diff
@@ -0,0 +1,77 @@
+// @flow
+
+import React, { Component } from 'react';
+import { parse, stringify } from 'qs';
+
+import type { ConfigOptions, PaginationProps } from './pagination.types';
+import type { ComponentType } from 'react';
+
+export const connectRouterWithPagination = (
+  RenderComponent: ComponentType<any>,
+  {
+    currentPageKey = 'currentPage',
+    pageChangeCallbackKey = 'onChange',
+    // above two values are use to set names of props, which are passing directly to pagination render component
+    pageParamName = 'page', // determines what is the name of page param, which is displaying in URL
+  }: ConfigOptions,
+) => (WrappedComponent: ComponentType<any>) => {
+  return class extends Component<PaginationProps> {
+    // update history with new page param
+    handlePageChange = (page: number) => {
+      const queryParams = this.parseQueryParamsToObject();
+      queryParams[pageParamName] = page;
+      this.props.history.push(this.props.location.pathname + '?' + stringify(queryParams));
+    };
+
+    parseQueryParamsToObject = () => parse(this.props.location.search.slice(1));
+
+    setDefaultPageParam = (params: { [string]: string | number }) => {
+      if (!params[pageParamName]) {
+        params[pageParamName] = 1;
+        const paramsString = stringify(params);
+        window.history.pushState({}, null, window.location.origin + this.props.location.pathname + '?' + paramsString);
+      }
+    };
+
+    componentDidMount() {
+      const queryParams = this.parseQueryParamsToObject();
+      // if page param is missing in URL set its value to 1
+      if (!queryParams[pageParamName]) {
+        this.setDefaultPageParam(queryParams);
+      }
+      //dispatching redux action
+      this.props.onPageChange({
+        page: queryParams[pageParamName],
+      });
+    }
+
+    componentDidUpdate(prevProps: PaginationProps) {
+      const queryParams = this.parseQueryParamsToObject();
+      // if page param is missing in URL set its value to 1
+      if (!queryParams[pageParamName]) {
+        this.setDefaultPageParam(queryParams);
+      }
+      if (prevProps.location.search !== this.props.location.search) {
+        //dispatching redux action
+        this.props.onPageChange({
```

Comment:
It should be called in `handlePageChange `, there is no need to check if queryParams has changed